File: frontend/lib/data/models/report.dart
import 'enums.dart';
import 'material_info.dart' show ApiErrorInfo;

/// 단일 세션 리포트 (GET /sessions/{id}/report, spec §5.2).
class Report {
  const Report({
    required this.status,
    this.typeScores = const {},
    this.strongTypes = const [],
    this.weakTypes = const [],
    this.speakingHabits,
    this.insight,
    this.error,
  });

  final AsyncStatus status;

  /// QuestionStrategy별 답변 점수 0.0~1.0 — 성장 리포트의 원천.
  final Map<QuestionStrategy, double> typeScores;
  final List<QuestionStrategy> strongTypes;
  final List<QuestionStrategy> weakTypes;
  final SpeakingHabits? speakingHabits;
  final String? insight;
  final ApiErrorInfo? error;

  factory Report.fromJson(Map<String, dynamic> json) {
    final scores = <QuestionStrategy, double>{};
    (json['type_scores'] as Map<String, dynamic>? ?? {}).forEach((k, v) {
      scores[QuestionStrategy.fromWire(k)] = (v as num).toDouble();
    });
    final quality = json['answer_quality'] as Map<String, dynamic>? ?? {};
    List<QuestionStrategy> parseTypes(String key) =>
        (quality[key] as List<dynamic>? ?? [])
            .map((e) => QuestionStrategy.fromWire(e as String))
            .toList();

    return Report(
      status: AsyncStatus.fromWire(json['status'] as String),
      typeScores: scores,
      strongTypes: parseTypes('strong_types'),
      weakTypes: parseTypes('weak_types'),
      speakingHabits: json['speaking_habits'] == null
          ? null
          : SpeakingHabits.fromJson(
              json['speaking_habits'] as Map<String, dynamic>),
      insight: json['insight'] as String?,
      error: json['error'] == null
          ? null
          : ApiErrorInfo.fromJson(json['error'] as Map<String, dynamic>),
    );
  }
}

/// 발표 습관 — WPM은 필러 포함(팀 협의 확정), over_time은 클라이언트 파생.
class SpeakingHabits {
  const SpeakingHabits({
    required this.wordsPerMinute,
    this.fillerWords = const [],
    required this.timeLimitSeconds,
    required this.actualSeconds,
  });

  final double wordsPerMinute;
  final List<FillerWord> fillerWords;
  final int timeLimitSeconds;
  final int actualSeconds;

  int get fillerWordCount =>
      fillerWords.fold(0, (sum, f) => sum + f.count);
  bool get overTime => actualSeconds > timeLimitSeconds;

  factory SpeakingHabits.fromJson(Map<String, dynamic> json) => SpeakingHabits(
        wordsPerMinute: (json['words_per_minute'] as num).toDouble(),
        fillerWords: (json['filler_words'] as List<dynamic>? ?? [])
            .map((e) => FillerWord.fromJson(e as Map<String, dynamic>))
            .toList(),
        timeLimitSeconds: json['time_limit_seconds'] as int,
        actualSeconds: json['actual_seconds'] as int,
      );
}

class FillerWord {
  const FillerWord({required this.word, required this.count});
  final String word;
  final int count;

  factory FillerWord.fromJson(Map<String, dynamic> json) =>
      FillerWord(word: json['word'] as String, count: json['count'] as int);
}

/// 성장 리포트 (GET /users/me/report/growth — 유저 스코프, spec E).
class GrowthReport {
  const GrowthReport({
    required this.range,
    this.teamId,
    this.series = const [],
    this.insight,
  });

  final String range;
  final String? teamId;
  final List<GrowthPoint> series;
  final String? insight;

  factory GrowthReport.fromJson(Map<String, dynamic> json) => GrowthReport(
        range: json['range'] as String,
        teamId: json['team_id'] as String?,
        series: (json['series'] as List<dynamic>? ?? [])
            .map((e) => GrowthPoint.fromJson(e as Map<String, dynamic>))
            .toList(),
        insight: json['insight'] as String?,
      );
}

class GrowthPoint {
  const GrowthPoint({
    required this.sessionId,
    required this.name,
    required this.date,
    this.typeScores = const {},
  });

  final String sessionId;
  final String name;
  final String date;
  final Map<QuestionStrategy, double> typeScores;

  factory GrowthPoint.fromJson(Map<String, dynamic> json) {
    final scores = <QuestionStrategy, double>{};
    (json['type_scores'] as Map<String, dynamic>? ?? {}).forEach((k, v) {
      scores[QuestionStrategy.fromWire(k)] = (v as num).toDouble();
    });
    return GrowthPoint(
      sessionId: json['session_id'] as String,
      name: json['name'] as String,
      date: json['date'] as String,
      typeScores: scores,
    );
  }
}

File: backend/app/schemas/report.py
from pydantic import BaseModel, Field, field_validator

from app.db.enums import AsyncStatus, QuestionStrategy
from app.schemas.session import ErrorInfo


class ReportDraft(BaseModel):
    """리포트 정성 평가 1건 — LLM 산출(report-eval-workflow.md §B).

    팀원2 라우터가 `type_scores` → report_type_scores 행들, `insight` → reports.insight로 저장.
    정량 지표(WPM·filler·over_time)는 services/report.py(§A)에서 별도 계산 — 여기 없음.
    """

    # 전략별 답변 점수 0.0~1.0 (성장 리포트의 원천, api-spec §5.2).
    # 답변이 등장한 전략만 포함(0개인 전략 키는 생략).
    type_scores: dict[QuestionStrategy, float] = Field(default_factory=dict)
    insight: str = ""

    @field_validator("type_scores")
    @classmethod
    def _clamp_scores(cls, v: dict[QuestionStrategy, float]) -> dict[QuestionStrategy, float]:
        # DDL CHECK(score BETWEEN 0 AND 1) 위반 방어 — 범위 밖 값은 잘라낸다.
        return {k: max(0.0, min(1.0, float(s))) for k, s in v.items()}


# ── 응답 (리포트 라우트, api-spec §5.2) ────────────────────────────────
# time_limit_seconds·actual_seconds·answer_quality는 저장 컬럼이 없고 서빙 시
# 파생(db-schema §5). over_time·filler_word_count는 클라이언트 파생이라 응답에 없음.


class FillerWordOut(BaseModel):
    word: str
    count: int


class SpeakingHabitsOut(BaseModel):
    """발표 습관 — 수치는 services/report.py(§A) 저장값 + sessions·recordings 파생."""

    words_per_minute: float
    filler_words: list[FillerWordOut]
    time_limit_seconds: int   # sessions.time_limit_minutes * 60
    actual_seconds: int       # recordings.duration_seconds


class AnswerQualityOut(BaseModel):
    """type_scores 임계값 분류 — 저장하지 않고 응답 시 파생 (§5.2)."""

    strong_types: list[QuestionStrategy]
    weak_types: list[QuestionStrategy]


class ReportOut(BaseModel):
    """GET /sessions/{id}/report 응답 (§5.2). ready 전에는 status(+error)만 채워진다."""

    status: AsyncStatus
    type_scores: dict[QuestionStrategy, float] | None = None
    answer_quality: AnswerQualityOut | None = None
    speaking_habits: SpeakingHabitsOut | None = None
    insight: str | None = None
    error: ErrorInfo | None = None   # failed 시 {code, message}


class GrowthPointOut(BaseModel):
    """성장 리포트 시리즈 1점 = 완료 세션 1개의 전략별 점수."""

    session_id: str
    name: str
    date: str                        # "YYYY-MM-DD" (ended_at 없으면 created_at)
    type_scores: dict[QuestionStrategy, float]


class GrowthReportOut(BaseModel):
    """GET /users/me/report/growth 응답 (§5.2, E: 유저 스코프)."""

    range: str
    user_id: str
    team_id: str | None = None
    series: list[GrowthPointOut]
    insight: str | None = None
